Content.Server/Radio/EntitySystems/MessagesServerSystem.cs

```diff
@@ -0,0 +1,148 @@
+
+                Sync(uid, server);
+            }
+        }
+    }
+
+    public void Update(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+
+        if (!TryComp(uid, out ApcPowerReceiverComponent? powerReceiver) || !(powerReceiver.Powered))
+            return;
+
+        var query = EntityQueryEnumerator<MessagesCartridgeComponent>();
+        List<(string,string)> toUpdate = new();
+
+        bool needDictUpdate = false;
+
+        while(query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if ((cartComponent.UserUid == null) || (cartComponent.UserName == null))
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+            if ((component.NameDict.ContainsKey(cartComponent.UserUid)) && (component.NameDict[cartComponent.UserUid] == cartComponent.UserName))
+                continue;
+
+            needDictUpdate = true;
+            component.NameDict[cartComponent.UserUid] = cartComponent.UserName;
+            toUpdate.Add((cartComponent.UserUid,cartComponent.UserName));
+        }
+
+        if (needDictUpdate)
+        {
+            query = EntityQueryEnumerator<MessagesCartridgeComponent>();
+            while (query.MoveNext(out var cartUid,out var cartComponent))
+            {
+                foreach (var (key,value) in toUpdate)
+                {
+                    cartComponent.NameDict[key]=value;
+                }
+            }
+        }
+
+        if ((component.MessagesQueue.Count > 0))
+        {
+            var tempMessagesQueue = new List<MessagesMessageData>(component.MessagesQueue);
+            foreach (var message in tempMessagesQueue)
+            {
+                bool sent=TryToSend(message, mapId, component);
+                if (sent)
+                {
+                    component.MessagesQueue.Remove(message);
+                    component.Messages.Add(message);
+                }
+            }
+        }
+    }
+
+    public void Sync(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+        var query = EntityQueryEnumerator<MessagesCartridgeComponent>();
+        while(query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if ((cartComponent.UserUid == null) || (cartComponent.UserName == null))
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+
+            foreach (var entry in component.NameDict)
+            {
+                if (cartComponent.NameDict[entry.Key] != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+        }
+    }
+
```

Comment:
Having polling in both PDAs and server is unnecessary, you should just be doing it on the server to keep the code significantly cleaner than this and the copied checks everywhere.

Additionally this code must be able to handle generic messengers (for example PAIs) and not just be hardcoded to PDAs.

Reply:
See my comment below on:
- The caching handling looks really dodgy.
- You should be storing what components are valid up-front and not re-iterating it every time for TrySend.